File: Golang/easy/Exercism/need-for-speed/need_for_speed.go
package speed

type Car struct {
	battery      int
	batteryDrain int
	speed        int
	distance     int
}

type Track struct {
	distance int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	car := Car{
		battery:      100,
		batteryDrain: batteryDrain,
		speed:        speed,
	}
	return car
}

// NewTrack created a new track
func NewTrack(distance int) Track {
	track := Track{distance: distance}
	return track
}

// Drive drives the car one time. If there is not enough battery to drive on more time,
// the car will not move.
func Drive(car Car) Car {
	if car.battery >= car.batteryDrain {
		car.battery = car.battery - car.batteryDrain
		car.distance = car.distance + car.speed
	}
	return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
	return (car.battery/car.batteryDrain)*car.speed >= track.distance
}

File: Java/easy/Exercism/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int battery = 100;
    private int speed;
    private int distanceDriven = 0;
    private int drained;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.drained = batteryDrain;

    }

    public boolean batteryDrained() {
        return battery == 0 ? true:false;
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
        if(battery >= this.drained){
            this.distanceDriven += this.speed;
            this.battery -= drained;
        }

    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;
    public RaceTrack(int distance){
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        while(!car.batteryDrained()){
            car.drive();
        }

        return car.distanceDriven() >= distance;
    }
}
